File: src-tauri/src/commands.rs
use crate::{autostart, config, config::Config, state::AppState, toast, translate, tray};
use tauri::{AppHandle, State};

use std::fs;

type CommandResult<T> = Result<T, String>;

#[tauri::command]
pub fn get_config(state: State<'_, AppState>) -> Config {
    state.config()
}

#[tauri::command]
pub fn set_config(cfg: Config, state: State<'_, AppState>) -> CommandResult<()> {
    config::save_config(&cfg).map_err(|error| error.to_string())?;
    state.set_config(cfg);
    Ok(())
}

#[tauri::command]
pub fn pause_caps_lock(
    paused: bool,
    app: AppHandle,
    state: State<'_, AppState>,
) -> CommandResult<()> {
    state.set_caps_paused(paused);
    tray::rebuild_tray_menu(&app).map_err(|error| error.to_string())
}

#[tauri::command]
pub fn is_caps_paused(state: State<'_, AppState>) -> bool {
    state.caps_paused()
}

#[tauri::command]
pub fn set_auto_start(enabled: bool) -> CommandResult<()> {
    autostart::set_auto_start(enabled).map_err(|error| error.to_string())
}

#[tauri::command]
pub fn is_auto_start() -> CommandResult<bool> {
    autostart::is_auto_start().map_err(|error| error.to_string())
}

#[tauri::command]
pub fn test_translate_api(
    server_url: String,
    api_key: String,
    target: String,
) -> CommandResult<String> {
    translate::test_translate_api(&server_url, &api_key, &target)
}

#[tauri::command]
pub fn replace_with_translation(text: String) -> CommandResult<()> {
    translate::replace_with_translation(text)
}

#[tauri::command]
pub fn copy_to_clipboard(text: String) -> CommandResult<()> {
    translate::copy_to_clipboard(text)
}

#[tauri::command]
pub fn hide_translation_toast() {
    toast::hide_translation_toast();
}

#[tauri::command]
pub fn export_replacements_json(json: String) -> CommandResult<bool> {
    let dialog = rfd::FileDialog::new()
        .set_title("Экспорт замен")
        .set_file_name("keytweak-replacements.json")
        .add_filter("JSON", &["json"]);

    match dialog.save_file() {
        Some(path) => {
            fs::write(&path, json.as_bytes()).map_err(|e| e.to_string())?;
            Ok(true)
        }
        None => Ok(false), // user cancelled
    }
}

#[tauri::command]
pub fn import_replacements_json() -> CommandResult<Option<String>> {
    let dialog = rfd::FileDialog::new()
        .set_title("Импорт замен")
        .add_filter("JSON", &["json"]);

    match dialog.pick_file() {
        Some(path) => {
            let content = fs::read_to_string(&path).map_err(|e| e.to_string())?;
            Ok(Some(content))
        }
        None => Ok(None), // user cancelled
    }
}

File: src/lib/commands.ts
import { invoke } from "@tauri-apps/api/core";
import type { Config } from "../types/config";

declare global {
  interface Window {
    __TAURI_INTERNALS__?: unknown;
  }
}

const defaultConfig: Config = {
  caps_lock: {
    switch_mode: "previous",
    switch_key: "capslock",
    real_caps_combo: "shift_caps",
    auto_start: false,
    paused: false,
  },
  auto_replace: {
    trigger_space: true,
    trigger_tab: true,
    trigger_enter: false,
    trigger_punctuation: true,
    whole_words_only: true,
    case_sensitive: false,
    replacements: [],
  },
  key_remap: {
    enabled: true,
    mappings: [{ from: "left_alt", to: "win", enabled: false }],
  },
  translate: {
    server_url: "http://127.0.0.1:5000",
    api_key: "",
    auto_detect_language: true,
    target_language: "ru",
    hotkey_translate: "ctrl+c+c",
    hotkey_reverse: "ctrl+shift+c",
  },
  general: {
    app_language: "en",
    theme: "system",
  },
  exception_mode: "blacklist",
  exceptions: [],
};

function inTauri() {
  return typeof window !== "undefined" && Boolean(window.__TAURI_INTERNALS__);
}

function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error(`IPC call timed out after ${ms}ms`)),
      ms,
    );
    promise.then(
      (value) => {
        clearTimeout(timer);
        resolve(value);
      },
      (error) => {
        clearTimeout(timer);
        reject(error);
      },
    );
  });
}

export function getConfig() {
  if (!inTauri()) return Promise.resolve(defaultConfig);
  return withTimeout(invoke<Config>("get_config"), 10000);
}

export function setConfig(cfg: Config) {
  if (!inTauri()) return Promise.resolve(void cfg);
  return invoke<void>("set_config", { cfg });
}

export function pauseCapsLock(paused: boolean) {
  if (!inTauri()) return Promise.resolve(void paused);
  return invoke<void>("pause_caps_lock", { paused });
}

export function isCapsPaused() {
  if (!inTauri()) return Promise.resolve(false);
  return invoke<boolean>("is_caps_paused");
}

export function setAutoStart(enabled: boolean) {
  if (!inTauri()) return Promise.resolve(void enabled);
  return invoke<void>("set_auto_start", { enabled });
}

export function isAutoStart() {
  if (!inTauri()) return Promise.resolve(defaultConfig.caps_lock.auto_start);
  return invoke<boolean>("is_auto_start");
}

export function testTranslateApi(
  serverUrl: string,
  apiKey: string,
  target: string,
) {
  if (!inTauri()) return Promise.resolve(target === "ru" ? "привет" : "hello");
  return invoke<string>("test_translate_api", { serverUrl, apiKey, target });
}

export function replaceWithTranslation(text: string) {
  if (!inTauri()) return Promise.resolve();
  return invoke<void>("replace_with_translation", { text });
}

export function copyToClipboard(text: string) {
  if (!inTauri())
    return navigator.clipboard?.writeText(text) ?? Promise.resolve();
  return invoke<void>("copy_to_clipboard", { text });
}

export function hideTranslationToast() {
  if (!inTauri()) return Promise.resolve();
  return invoke<void>("hide_translation_toast");
}

export function exportReplacementsJson(json: string) {
  if (!inTauri()) return Promise.resolve(false);
  return invoke<boolean>("export_replacements_json", { json });
}

export function importReplacementsJson() {
  if (!inTauri()) return Promise.resolve(null as string | null);
  return invoke<string | null>("import_replacements_json");
}

export interface RunningProgram {
  exe_name: string;
  display_name: string;
}

export function getRunningPrograms() {
  if (!inTauri()) return Promise.resolve([] as RunningProgram[]);
  return invoke<RunningProgram[]>("get_running_programs");
}

export function pickProgramFile() {
  if (!inTauri()) return Promise.resolve(null as string | null);
  return invoke<string | null>("pick_program_file");
}
